src/dex/swaap-v2/swaap-v2.ts

```diff
@@ -0,0 +1,713 @@
+      this.dexKey,
+      normalizedSrcToken.address,
+      normalizedDestToken.address,
+    );
+
+    if (normalizedSrcToken.address === normalizedDestToken.address) {
+      return null;
+    }
+
+    const pools =
+      limitPools ??
+      (await this.getPoolIdentifiers(srcToken, destToken, side, blockNumber));
+
+    const levels = await this.getCachedLevels();
+    if (levels === null) {
+      return null;
+    }
+
+    const levelEntries: SwaapV2PriceLevels[] = Object.keys(levels)
+      .map((pair: string) => {
+        if (
+          pools.includes(
+            getPoolIdentifier(
+              this.dexKey,
+              normalizedSrcToken.address,
+              normalizedDestToken.address,
+            ),
+          )
+        ) {
+          return levels[pair];
+        }
+        return undefined;
+      })
+      .filter(o => o !== undefined)
+      .map(o => o!);
+
+    const prices = levelEntries.map((askAndBids: SwaapV2PriceLevels) => {
+      const divider = getBigNumberPow(
+        (side === SwapSide.SELL ? normalizedSrcToken : normalizedDestToken)
+          .decimals,
+      );
+
+      const amountsFloat = amounts.map(a =>
+        new BigNumber(a.toString()).dividedBy(divider),
+      );
+
+      const unitPrice = this.computePricesFromLevelsBis(
+        [BN_1],
+        askAndBids,
+        normalizedSrcToken,
+        normalizedDestToken,
+        side,
+      )[0];
+      const prices = this.computePricesFromLevelsBis(
+        amountsFloat,
+        askAndBids,
+        normalizedSrcToken,
+        normalizedDestToken,
+        side,
+      );
+
+      return {
+        gasCost: GAS_COST_ESTIMATION,
+        exchange: this.dexKey,
+        data: {},
+        prices,
+        unit: unitPrice,
+        poolIdentifier: requestedPoolIdentifier,
+        poolAddresses: [this.routerAddressPlaceholder],
+      } as PoolPrices<SwaapV2Data>;
+    });
+
+    return prices;
+  }
+
+  getBaseToken(poolIdentifier: string): string {
+    return poolIdentifier.split('_')[1];
+  }
+
+  getLevels(
+    askAndBids: SwaapV2PriceLevels,
+    side: SwapSide,
+    poolIdentifier: string,
+  ) {
+    const baseToken: string = this.getBaseToken(poolIdentifier);
+    if (side === SwapSide.SELL) {
+      return askAndBids.asks;
+    }
+    return askAndBids.asks;
+  }
+
+  async preProcessTransaction(
+    optimalSwapExchange: OptimalSwapExchange<SwaapV2Data>,
+    srcToken: Token,
+    destToken: Token,
+    side: SwapSide,
+    options: PreprocessTransactionOptions,
+  ): Promise<[OptimalSwapExchange<SwaapV2Data>, ExchangeTxInfo]> {
+    const isSell = side === SwapSide.SELL;
+
+    const normalizedSrcToken = this.normalizeToken(srcToken);
+    const normalizedDestToken = this.normalizeToken(destToken);
+
+    const tolerance = (
+      options.slippageFactor > BN_1
+        ? options.slippageFactor.minus(BN_1)
+        : BN_1.minus(options.slippageFactor)
+    ).toNumber();
+
+    const quote = await this.rateFetcher.getQuote(
```

Comment:
I don't see restriction mechanism in case of failure:
https://github.com/paraswap/paraswap-dex-lib/blob/0bb956f7c447531fd1ac21ff9bec57ade7d13936/src/dex/hashflow/hashflow.ts#L696-L710
like we have for Hashflow

Reply:
Adding that we should also have our own slippage check, similar to [hashflow](https://github.com/paraswap/paraswap-dex-lib/blob/a4669eb522a3b0805d676fe905aa2e0502e62a7d/src/dex/hashflow/hashflow.ts#L631) or [generic rfq](https://github.com/paraswap/paraswap-dex-lib/blob/c7a2f19dbafb0f2357927c7d2f0c21a948c1bfa4/src/dex/generic-rfq/generic-rfq.ts#L225)